# DISA Database design

The first task in recreating the DISA database was migrating data from the existing Mongo db. First, let's take a look at a sample record from the Mongo db (**disamg**), the contents of which have been dumped into a file *latest_entries.json*. 

In [1]:
import json
from collections import namedtuple
import datetime as dt
import pandas as pd

with open('data/latest-entries.json') as f:
    data = json.load(f)

In [2]:
data[0]

{'_id': {'$oid': '598b4c8baecb6c0001a29e25'},
 'person': {'names': [{'firstName': 'Elizabeth',
    'lastName': '',
    'type': 'English'}],
  'tribe': 'Unspecified',
  'origin': '',
  'race': 'Indian',
  'sex': 'Female',
  'typeKindOfEnslavement': 'Servant',
  'vocation': '',
  'father': {'name': {'firstName': '', 'lastName': ''},
   'race': '',
   'origin': '',
   'status': '',
   'owner': {'name': {'firstName': '', 'lastName': '', 'title': ''},
    'vocation': ''}},
  'mother': {'name': {'firstName': '', 'lastName': ''},
   'race': '',
   'origin': '',
   'status': '',
   'owner': {'name': {'firstName': '', 'lastName': '', 'title': ''},
    'vocation': ''}},
  'children': []},
 'owner': {'name': {'firstName': '', 'lastName': 'Duglas', 'title': 'Mr. '},
  'vocation': ''},
 'dateOfRunaway': {'month': ''},
 'dateOfEmancipation': {'month': ''},
 'buyer': {'firstName': '', 'lastName': ''},
 'seller': {'firstName': '', 'lastName': ''},
 'dateOfSale': {'month': ''},
 'dateOfMarriage': {'mon

We see a number of potential issues here. Let's consider each in turn.

### MongoDB Document Structure

Each document in any Mongo db is a JSON object, which itself can contain an arbitrary number of JSON objects. Here are the top-level attributes for all documents in **disamg**:

In [3]:
all_attrs = { k for d in data for k in d.keys() }
for a in sorted(list(all_attrs)):
    print(a)

_id
additionalInformation
buyer
dateOfDeath
dateOfEmancipation
dateOfMarriage
dateOfRunaway
dateOfSale
document
meta
owner
person
researcherNotes
seller


This is an odd assortment of labels; it's not immediately clear what each document in **disamg** represents. Considering the context of the DISA project, the guess is that `person` is the central data point, with `buyer`, `seller`, `owner`, and the 5 `dateOf*` fields referring back to `person` (ie, 'owner of `person`, 'date of death of `person`, etc.). `document` presumably identifies the historical document where the `person` appears, with `_id`, `meta`, `additionalInformation`, and `researcherNotes` all containing some administrative-level information.

The attributes `researcherNotes` and `additionalInformation` are both text strings; all the rest of the attributes are themselves JSON objects containing more data. We'll take a poll of these object-attributes, and see what fields are available in them. One of the challenges with this or any other Mongo db is that the data structure is potentuially inconsistent. Then, we'll see how much of this data is empty, as the sample document shows several null values.

In [5]:
str_vals = { 'additionalInformation', 'researcherNotes' }
obj_vals = { 'buyer', 'dateOfDeath', 'dateOfEmancipation', 'dateOfMarriage', 'dateOfRunaway',
            'dateOfSale', 'document', 'meta', 'owner', 'person', 'seller', '_id' }

from collections import Counter
count_attrs = { a: Counter() for a in all_attrs if a not in str_vals }
count_attrs['researcherNotes'] = 0
count_attrs['additionalInformation'] = 0
total_docs = 0
for d in data:
    total_docs += 1
    for attr in all_attrs:
        v = d.get(attr)
        if attr in str_vals and v is not None:
            count_attrs[attr] += 1
        elif v:
            for n in v.keys():
                count_attrs[attr][n] += 1

print('Total Mongo Documents: {}\n'.format(total_docs))
print('Document attribute counts\n')
for a in sorted(list(count_attrs.keys())):
    print(a)
    if a in obj_vals:
        for v, c in count_attrs[a].items():
            print('\t{}:{}'.format(v, c))
    else:
        print('\t',count_attrs[a])
    print()

Total Mongo Documents: 493

Document attribute counts

_id
	$oid:493

additionalInformation
	 493

buyer
	firstName:493
	lastName:493

dateOfDeath
	month:353
	year:1
	day:1

dateOfEmancipation
	month:493
	year:31
	day:32

dateOfMarriage
	month:493
	year:2
	day:2

dateOfRunaway
	month:493
	year:68
	day:58

dateOfSale
	month:493
	year:1
	day:1

document
	date:493
	stringLocation:493
	colonyState:493
	nationalContext:493
	sourceType:493
	recordType:493
	citation:493
	locale:3

meta
	stage:493
	creator:493
	lastModified:493
	updatedBy:493
	idPrefix:493
	idSuffix:493

owner
	name:493
	vocation:493

person
	names:493
	tribe:493
	origin:493
	race:493
	sex:493
	typeKindOfEnslavement:493
	vocation:493
	father:493
	mother:493
	children:493
	age:405

researcherNotes
	 493

seller
	firstName:493
	lastName:493



The attribute counts look reasonably consistent with the total record count. The biggest discrepencies are in the `dateOf*` objects. Looking back at the sample record, we see that an empty date is stored as `{'month': ''}`; and that probably accounts for why there is a difference between `month`, `year`, and `day` counts for individual `dateOf` attributes.

We'll guess at empty data structures for other attribute values, and see just how much this data is empty. We'll gloss the `person` attribute for now, since that will require a more involved treatment. Again reveiwing the sample record, we see that several of the `person` attributes -- `names`, `mother`, `father`,  and `children` -- are themselves more nested JSON data.

In [25]:
empty_attr = {
    '_id' : [{'$oid': ''}],
    'additionalInformation': [''],
    'buyer': [{'firstName': '', 'lastName': ''}],
    'dateOfRunaway': [{'month': ''}],
    'dateOfSale': [{'month': ''}],
    'dateOfMarriage': [{'month': ''}],
    'dateOfEmancipation': [{'month': ''}],
    'dateOfDeath': [{'month': ''}],
    'document': [{'date': {'month': ''}, 'stringLocation': '', 'colonyState': '', 'nationalContext': '',
                'sourceType': '', 'recordType': '', 'citation': ''},
                {'date': {'month': ''}, 'stringLocation': '', 'colonyState': '', 'nationalContext': '',
                'sourceType': '', 'recordType': '', 'citation': '', 'locale': ''}],
    'meta': [{'stage': '', 'creator': '', 'lastModified': '', 'updatedBy': '', 'idPrefix': '', 'idSuffix': ''}],
    'owner': [{'name': {'firstName': '', 'lastName': '', 'title': ''}, 'vocation': ''}],
    'researcherNotes': [''],
    'seller': [{'firstName': '', 'lastName': ''}]
}

count_empty = { k: 0 for k in empty_attr.keys() }
for d in data:
    for e, checks in empty_attr.items():
        v = d.get(e)
        if not v:
            count_empty[e] += 1
            continue
        for c in checks:
            if v == c:
                count_empty[e] += 1

print('Empty attribute counts\n')
for a in sorted(list(count_empty.keys())):
    print('{0}: {1} empty ({2:.2f}% present)'.format(a, count_empty[a], 100-(count_empty[a]/total_docs)*100))

Empty attribute counts

_id: 0 empty (100.00% present)
additionalInformation: 109 empty (77.89% present)
buyer: 493 empty (0.00% present)
dateOfDeath: 492 empty (0.20% present)
dateOfEmancipation: 461 empty (6.49% present)
dateOfMarriage: 491 empty (0.41% present)
dateOfRunaway: 425 empty (13.79% present)
dateOfSale: 492 empty (0.20% present)
document: 3 empty (99.39% present)
meta: 0 empty (100.00% present)
owner: 75 empty (84.79% present)
researcherNotes: 381 empty (22.72% present)
seller: 486 empty (1.42% present)


Some quick analysis:
* `buyer`, `seller`, `dateOfSale`, `dateOfMarriage`, `dateOfDeath` are almost entirely empty
* `owner` and `additionalInformation` are about 80% present
* `researcherNotes`, `dateOfEmancipation` and `dateOfRunaway` are less common, but still occasionally filled out
* `_id` and `meta` are present on all **disamg** documents

What about those 3 empty `document` values? Let's examine:

In [26]:
for d in data:
    if d['document'] in empty_attr['document']:
        for k,v in d.items():
            print('{}:\n\t{}'.format(k,v))
        print('='*100)

_id:
	{'$oid': '59b98bebdf4c6f000198331e'}
person:
	{'names': [{'firstName': 'FIRST NAME', 'lastName': '', 'type': 'English'}, {'firstName': '', 'lastName': 'LAST NAME', 'type': 'Indian'}], 'tribe': '', 'origin': '', 'race': '', 'sex': '', 'typeKindOfEnslavement': '', 'vocation': '', 'father': {'name': {'firstName': '', 'lastName': ''}, 'race': '', 'origin': '', 'status': '', 'owner': {'name': {'firstName': '', 'lastName': '', 'title': ''}, 'vocation': ''}}, 'mother': {'name': {'firstName': '', 'lastName': ''}, 'race': '', 'origin': '', 'status': '', 'owner': {'name': {'firstName': '', 'lastName': '', 'title': ''}, 'vocation': ''}}, 'children': []}
owner:
	{'name': {'firstName': '', 'lastName': '', 'title': ''}, 'vocation': ''}
dateOfRunaway:
	{'month': ''}
dateOfEmancipation:
	{'month': ''}
buyer:
	{'firstName': '', 'lastName': ''}
seller:
	{'firstName': '', 'lastName': ''}
dateOfSale:
	{'month': ''}
dateOfMarriage:
	{'month': ''}
document:
	{'date': {'month': ''}, 'stringLocation': '

Two of these appear to be test documents, meaning there is only 1 of the 493 without `document` data. We'll count this as an error.

If we were desiging an entity-relationship diagram for **disamg** as-is, it might look something like this:

![title](img/DISA_er_1.png)

We'll consider all nested object-attributes as independent entities for now, and the `additionalInformation` and `researcherNotes` fields as as entity attributes attendant on the Mongo document (whatever that might represent). A document has a mandatory relationship with `_id`, `meta`, `document`, and `person`, since these are present on nearly every **disamg** document. This may be a many-to-one relationship: many **disamg** documents might be associated with the same historical `document`, for example. For now, we can be sure that each **disamg** document is associated with 1 and only 1 of these entities, since none of the attributes are multivalued.

Relationships between a document and the remaining entities -- `buyer`, `seller`, `owner`, and `dateOf*` -- are optional, given by their inconsistent presence in our analysis. These too are unary relationships, since again none of them are multivalued. However, here that seems potentially problematic. If our hunch is correct -- that each **disamg** document stands in for an enslaved person -- we might expect `owner` or `dateOfSale` to be multivalued, since someone might be sold multiple times over their lifetime. Moreover, what's the relationship between `buyer`, `seller`, and `owner`? One might expect that the `buyer` and `seller` of a person would be an `owner` as well.

Before diving into those issues, let's examine the mandatory attributes `_id` and `meta`. These have fields that look like IDs, which may be helpful in determining just what a **disamg** document represents.

In [47]:
unique_data = {
    'id_oid': set(),
    'meta_id_prefix': set(),
    'meta_id_suffix': set(),
    'meta_id_stage': set(),
    'meta_creator': set(),
    'meta_updated_by': set(),
    'meta_modified': set(),
}

for d in data:
    unique_data['id_oid'].add(d['_id']['$oid'])
    unique_data['meta_id_prefix'].add(d['meta']['idPrefix'])
    unique_data['meta_id_suffix'].add(d['meta']['idSuffix'])
    unique_data['meta_id_stage'].add(d['meta']['stage'])
    unique_data['meta_creator'].add(d['meta']['creator'])
    unique_data['meta_updated_by'].add(d['meta']['updatedBy'])
    unique_data['meta_modified'].add(d['meta']['lastModified'])

for att, d in unique_data.items():
    d_list = list(d)
    print("Number of unique {{{}}} values: {}".format(att, len(d_list)))
    print("\tsamples: {}\n".format(d_list[:8]))

Number of unique {id_oid} values: 493
	samples: ['59cc7c8edf4c6f0001983343', '5a7094681b47910001a53ac6', '5a9497ff1b47910001a53b22', '5a6a03c91b47910001a53a93', '599f14a7aecb6c0001a29e45', '5b049339de60e051e7da37ff', '5a732dff1b47910001a53ae3', '5a6770dc1b47910001a53a76']

Number of unique {meta_id_prefix} values: 1
	samples: [10]

Number of unique {meta_id_suffix} values: 493
	samples: [10240, 10241, 10242, 10243, 10244, 10245, 10246, 10247]

Number of unique {meta_id_stage} values: 2
	samples: ['Draft', 'Internal']

Number of unique {meta_creator} values: 6
	samples: ['113112063790792171857', '109503335312366098524', '112487255676465508755', '106123372953260397156', '106895219236186746888', '112148132795694739523']

Number of unique {meta_updated_by} values: 8
	samples: ['113112063790792171857', '109503335312366098524', '112487255676465508755', '106123372953260397156', '106895219236186746888', '103795391716629952261', '112148132795694739523', '117289295548725522159']

Number of uniqu

In [ ]:
FlatRow = namedtuple('FlatRow', [
    #metadata
    'mongo_id', 'disa_id', 'modified_date', 'creator', 'last_updated_by',
    #name information
    'first_name', 'last_name', 'name_count',
    #personal details
    'tribe', 'origin', 'race', 'sex', 'age', 'enslavement', 'vocation',
    #personal event dates
    'date_runaway', 'date_sale', 'date_emancipation', 'date_marriage', 'date_death',
    #owner information
    'owner_first', 'owner_last', 'owner_title', 'owner_vocation',
    'buyer_first', 'buyer_last', 'seller_first', 'seller_last',
    #father information
    'father_first', 'father_last', 'father_origin', 'father_race', 'father_status',
    #father owner information
    'father_owner_first', 'father_owner_last', 'father_owner_title','father_owner_vocation',
    #mother information
    'mother_first', 'mother_last', 'mother_origin', 'mother_race', 'mother_status',
    #mother owner information
    'mother_owner_first', 'mother_owner_last', 'mother_owner_title', 'mother_owner_vocation',
    #child information
    'child_first', 'child_last', 'child_count',
    #location information
    'nation', 'state', 'location', 'locale',
    #document information
    'document_date', 'source', 'record', 'citation',
    #text blobs
    'addl_info', 'notes' ])

In [69]:
def get_date(dateData):
    if not dateData or dateData == {'month': ''} or dateData == {'year': '', 'month': '', 'day':''}:
        dateData = { 'day':1, 'month':1, 'year':1 }
    try:
        day = int( dateData.get('day',1) or 1)
        month = int( dateData.get('month',1) or 1)
        year = int( dateData.get('year',1) or 1)
    except:
        day = 1
        month = 1
        year = 1
    return dt.datetime(day=day, month=month, year=year) 

In [80]:
def nest(jd, *args):
    nest = jd
    for arg in args:
        try:
            nest = nest[arg]
        except:
            return ''
    return nest

def flatten(jd):
    row = FlatRow(
        mongo_id = nest(jd, '_id', '$oid'),
        disa_id = '{}-{}'.format(nest(jd,'meta','idPrefix'),nest(jd,'meta','idSuffix')),
        modified_date = dt.datetime.strptime(
            nest(jd, 'meta','lastModified'), '%Y-%m-%dT%H:%M:%S.%fZ'),
        creator = nest(jd, 'meta', 'creator'),
        last_updated_by = nest(jd, 'meta', 'updatedBy'),
        first_name = nest(jd, 'person', 'names', 0, 'firstName'),
        last_name = nest(jd, 'person', 'names', 0, 'firstName'),
        name_count = len( nest(jd, 'person', 'names') ),
        tribe = nest(jd, 'person', 'tribe'),
        origin = nest(jd, 'person', 'origin'),
        race = nest(jd, 'person', 'race'),
        sex = nest(jd, 'person', 'sex'),
        age = nest(jd, 'person', 'age'),
        enslavement = nest(jd, 'person', 'typeKindOfEnslavement'),
        vocation = nest(jd, 'person', 'vocation'),
        date_runaway = get_date( nest(jd, 'dateOfRunaway') ),
        date_sale = get_date( nest(jd, 'dateOfSale') ),
        date_emancipation = get_date( nest(jd, 'dateOfEmancipation') ),
        date_marriage = get_date( nest(jd, 'dateOfMarriage') ),
        date_death = get_date( nest(jd, 'dateOfDeath') ),
        owner_first = nest(jd, 'owner', 'firstName'),
        owner_last = nest(jd, 'owner', 'lastName'),
        owner_title = nest(jd, 'owner', 'title'),
        owner_vocation = nest(jd, 'owner', 'vocation'),
        buyer_first = nest(jd, 'buyer', 'firstName'),
        buyer_last = nest(jd, 'buyer', 'lastName'),
        seller_first = nest(jd, 'seller', 'firstName'),
        seller_last = nest(jd, 'seller', 'firstName'),
        father_first = nest(jd, 'person', 'father', 'name', 'firstName'),
        father_last = nest(jd, 'person', 'father', 'name', 'lastName'),
        father_origin = nest(jd, 'person', 'father', 'origin'),
        father_race = nest(jd, 'person', 'father', 'race'),
        father_status = nest(jd, 'person', 'father', 'status'),
        father_owner_first = nest(jd, 'person', 'father', 'owner', 'name', 'firstName'),
        father_owner_last = nest(jd, 'person', 'father', 'owner', 'name', 'lastName'),
        father_owner_title = nest(jd, 'person', 'father', 'owner', 'title'),
        father_owner_vocation = nest(jd, 'person', 'father', 'owner', 'vocation'),
        mother_first = nest(jd, 'person', 'mother', 'name', 'firstName'),
        mother_last = nest(jd, 'person', 'mother', 'name', 'lastName'),
        mother_origin = nest(jd, 'person', 'mother', 'origin'),
        mother_race = nest(jd, 'person', 'mother', 'race'),
        mother_status = nest(jd, 'person', 'mother', 'status'),
        mother_owner_first = nest(jd, 'person', 'mother', 'owner', 'name', 'firstName'),
        mother_owner_last = nest(jd, 'person', 'mother', 'owner', 'name', 'lastName'),
        mother_owner_title = nest(jd, 'person', 'mother', 'owner', 'title'),
        mother_owner_vocation = nest(jd, 'person', 'mother', 'owner', 'vocation'),
        child_first = nest(jd, 'person', 'children', 0, 'name', 'firstName'),
        child_last = nest(jd, 'person', 'children', 0, 'name', 'lastName'),
        child_count = len( nest(jd, 'person', 'children') ) ,
        nation = nest(jd, 'document', 'nationalContext'),
        state = nest(jd, 'document', 'colonyState'),
        location = nest(jd, 'document', 'stringLocation'),
        locale = nest(jd, 'document', 'locale'),
        document_date = get_date( nest(jd, 'document', 'date') ),
        source = nest(jd, 'document', 'sourceType'),
        record = nest(jd, 'document', 'recordType'),
        citation = nest(jd, 'document', 'citation'),
        addl_info = nest(jd, 'additionalInformation'),
        notes = nest(jd, 'researcherNotes') )

    return row

In [81]:
flat = [ flatten(rec) for rec in data ]
flat[0]

FlatRow(mongo_id='598b4c8baecb6c0001a29e25', disa_id='10-10002', modified_date=datetime.datetime(2018, 5, 14, 18, 32, 21, 539000), creator='112148132795694739523', last_updated_by='112487255676465508755', first_name='Elizabeth', last_name='Elizabeth', name_count=1, tribe='Unspecified', origin='', race='Indian', sex='Female', age='', enslavement='Servant', vocation='', date_runaway=datetime.datetime(1, 1, 1, 0, 0), date_sale=datetime.datetime(1, 1, 1, 0, 0), date_emancipation=datetime.datetime(1, 1, 1, 0, 0), date_marriage=datetime.datetime(1, 1, 1, 0, 0), date_death=datetime.datetime(1, 1, 1, 0, 0), owner_first='', owner_last='', owner_title='', owner_vocation='', buyer_first='', buyer_last='', seller_first='', seller_last='', father_first='', father_last='', father_origin='', father_race='', father_status='', father_owner_first='', father_owner_last='', father_owner_title='', father_owner_vocation='', mother_first='', mother_last='', mother_origin='', mother_race='', mother_status='',

In [82]:
disa = pd.DataFrame(flat)

In [83]:
disa

mongo_id   disa_id           modified_date  \
0    598b4c8baecb6c0001a29e25  10-10002 2018-05-14 18:32:21.539   
1    599c82feaecb6c0001a29e26  10-10003 2018-06-04 21:26:02.976   
2    599c8586aecb6c0001a29e27  10-10004 2018-05-22 22:58:51.373   
3    599c8922aecb6c0001a29e28  10-10005 2018-06-04 21:39:13.923   
4    599c8b4faecb6c0001a29e29  10-10006 2018-06-04 21:35:30.437   
..                        ...       ...                     ...   
488  5b0498c0de60e011b9da3802  10-10490 2018-05-23 04:57:46.281   
489  5b0499bbde60e00a7cda3803  10-10491 2018-05-23 04:58:07.227   
490  5b049b52de60e03fe7da3804  10-10492 2018-05-22 22:37:09.231   
491  5b049c9fde60e02bdada3805  10-10493 2018-05-23 17:23:40.210   
492  5b04a2f3de60e0606ada3806  10-10494 2018-06-04 21:40:53.256   

                   creator        last_updated_by first_name  last_name  \
0    112148132795694739523  112487255676465508755  Elizabeth  Elizabeth   
1    112148132795694739523  103795391716629952261      Harry      Harry   
2    112148132795694739523  109503335312366098524     Prince     Prince   
3    112148132795694739523  103795391716629952261     Hannah     Hannah   
4    112148132795694739523  103795391716629952261      Grace      Grace   
..                     ...                    ...        ...        ...   
488  109503335312366098524  109503335312366098524      Milly      Milly   
489  109503335312366098524  109503335312366098524                         
490  109503335312366098524  109503335312366098524                         
491  109503335312366098524  112148132795694739523                         
492  109503335312366098524  103795391716629952261                         

     name_count        tribe origin  ...   nation  \
0             1  Unspecified         ...  British   
1             1  Unspecified         ...  British   
2             1  Unspecified         ...  British   
3             1                      ...  British   
4             1                      ...  British   
..          ...          ...    ...  ...      ...   
488           1                      ...            
489           0                      ...            
490           1                      ...            
491           0                      ...  British   
492           0                      ...            

                           state    location          locale  \
0                    Connecticut  New London                   
1    Narragansett (Rhode Island)   Kingstown                   
2                  Massachusetts      Boston                   
3                  Massachusetts  Barnstable                   
4                       New York                               
..                           ...         ...             ...   
488                  Connecticut  Stonington                   
489                Massachusetts      Boston                   
490                                                            
491                Massachusetts      Boston  Sudbury Street   
492                Massachusetts      Boston                   

           document_date     source                 record  \
0    1671-05-07 00:00:00       Book                          
1    1704-06-12 00:00:00  Newspaper  Runaway advertisement   
2    1704-11-06 00:00:00  Newspaper  Runaway advertisement   
3    1706-02-17 00:00:00  Newspaper  Runaway advertisement   
4    1706-01-13 00:00:00  Newspaper  Runaway advertisement   
..                   ...        ...                    ...   
488  1716-08-20 00:00:00  Newspaper  Runaway Advertisement   
489  1716-08-27 00:00:00  Newspaper  Advertisement of sale   
490  0001-01-01 00:00:00                                     
491  1716-10-01 00:00:00  Newspaper  Advertisement of sale   
492  1706-04-22 00:00:00  Newspaper  Advertisement of sale   

                                              citation  \
0    Blake, The Later History of the First Church o...   
1    Zotero ID: DISA00004. America's Historical 

In [84]:
disa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 493 entries, 0 to 492
Data columns (total 59 columns):
mongo_id                 493 non-null object
disa_id                  493 non-null object
modified_date            493 non-null datetime64[ns]
creator                  493 non-null object
last_updated_by          493 non-null object
first_name               493 non-null object
last_name                493 non-null object
name_count               493 non-null int64
tribe                    493 non-null object
origin                   493 non-null object
race                     493 non-null object
sex                      493 non-null object
age                      493 non-null object
enslavement              493 non-null object
vocation                 493 non-null object
date_runaway             493 non-null object
date_sale                493 non-null object
date_emancipation        493 non-null object
date_marriage            493 non-null object
date_death               493 non-

In [86]:
disa[ disa.child_count != 0 ]

mongo_id   disa_id           modified_date  \
11   599cab56aecb6c0001a29e30  10-10013 2017-12-29 19:27:54.141   
22   599f04bdaecb6c0001a29e3b  10-10024 2017-12-29 19:36:48.169   
91   59ea38c1df4c6f000198334a  10-10093 2018-05-31 16:32:56.581   
92   5a0b84b6df4c6f000198334b  10-10094 2018-05-31 16:33:08.834   
119  5a3152addf4c6f0001983366  10-10121 2018-06-15 15:06:14.275   
120  5a316c27df4c6f0001983367  10-10122 2018-06-15 15:06:41.665   

                   creator        last_updated_by first_name last_name  \
11   113112063790792171857  112148132795694739523     Rachel    Rachel   
22   113112063790792171857  112148132795694739523       John      John   
91   112148132795694739523  113112063790792171857       Joan      Joan   
92   112148132795694739523  113112063790792171857   Catalina  Catalina   
119  112148132795694739523  113112063790792171857       Joan      Joan   
120  112148132795694739523  113112063790792171857       Joan      Joan   

     name_count        tribe              origin  ...    nation        state  \
11            1      Mohegan                      ...  British   Connecticut   
22            1  Unspecified                      ...   British      Jamaica   
91            1                  Pueblo de Pauto  ...   Spanish     Trinidad   
92            1               Pueblo de Casanare  ...   Spanish     Trinidad   
119           1                                   ...   Spanish     Trinidad   
120           1                         Casanave  ...   Spanish     Trinidad   

                location locale        document_date   source       record  \
11            New London         1720-12-04 00:00:00  Archive                
22            Port Royal         1749-04-27 00:00:00  Archive  Manumission   
91   San Joseph de Oruña         1688-07-22 00:00:00  Archive  Manumission   
92   San Joseph de Oruña         1688-07-22 00:00:00  Archive  Manumission   
119  San Joseph de Oruña         1688-07-22 00:00:00  Archive  Manumission   
120  San Joseph de Oruña         1688-07-22 00:00:00  Archive  Manumission   

                                              citation  \
11   Connecticut Church Records, Index, New London ...   
22   Jamaica Archives, Manumission of Slaves; “Manu...   
91   DISA00107; CARTAS DE GOBERNADORE, Archivo Gene...   
92   DISA00107; CARTAS DE GOBERNADORE, Archivo Gene...   
119  DISA00107; CARTAS DE GOBERNADORE, Archivo Gene...   
120  DISA00107; CARTAS DE GOBERNADORE, Archivo Gene...   

                              addl_info  \
11                                        
22                                        
91                                        
92                Married to Joan Pauto   
119                                       
120  Joan is noted as a widower (viudo)   

                                       notes  
11                                            
22       Same name between child and slave.   
91                                            
92   Husband's ID in database: DISA-10-10093  
119      Son's ID in database: DISA-10-10123  
120                                           

[6 rows x 59 columns]

In [87]:
disa.iloc[92].date_marriage

datetime.datetime(1, 1, 1, 0, 0)

In [90]:
disa[ disa.addl_info.str.contains("owner") ]

mongo_id   disa_id           modified_date  \
55   59c26e8edf4c6f0001983326  10-10057 2018-06-14 01:13:54.377   
126  5a383cc8df4c6f000198336d  10-10128 2018-06-15 15:18:44.544   
127  5a383e8ddf4c6f000198336e  10-10129 2018-06-15 15:18:58.994   
129  5a3841bfdf4c6f0001983370  10-10131 2018-06-15 15:19:56.478   
130  5a384276df4c6f0001983371  10-10132 2018-06-15 15:20:09.808   
..                        ...       ...                     ...   
429  5a949c181b47910001a53b2d  10-10431 2018-06-27 00:54:23.456   
430  5a949d191b47910001a53b2e  10-10432 2018-06-27 00:54:44.375   
431  5a949dca1b47910001a53b2f  10-10433 2018-06-27 00:54:53.798   
432  5a949e171b47910001a53b30  10-10434 2018-06-27 00:55:49.830   
433  5a949e581b47910001a53b31  10-10435 2018-06-27 00:56:10.394   

                   creator        last_updated_by first_name last_name  \
55   113112063790792171857  103795391716629952261     Israel    Israel   
126  113112063790792171857  113112063790792171857     Pickle    Pickle   
127  113112063790792171857  113112063790792171857     Johnny    Johnny   
129  113112063790792171857  113112063790792171857      Jacob     Jacob   
130  113112063790792171857  113112063790792171857      Catto     Catto   
..                     ...                    ...        ...       ...   
429  113112063790792171857  112148132795694739523   Mystilla  Mystilla   
430  113112063790792171857  112148132795694739523   Susannah  Susannah   
431  113112063790792171857  112148132795694739523      Quash     Quash   
432  113112063790792171857  112148132795694739523    Docinda   Docinda   
433  113112063790792171857  112148132795694739523    Patrona   Patrona   

     name_count        tribe         origin  ...   nation             state  \
55            1  Unspecified  Allentown, PA  ...  British     Pennsylvania    
126           1  Unspecified                 ...  British  British Honduras   
127           1  Unspecified                 ...  British  British Honduras   
129           1  Unspecified                 ...  British  British Honduras   
130           1  Unspecified                 ...  British  British Honduras   
..          ...          ...            ...  ...      ...               ...   
429           1  Unspecified                 ...  British  British Honduras   
430           1  Unspecified                 ...  British  British Honduras   
431           1  Unspecified                 ...  British  British Honduras   
432           1  Unspecified                 ...  British  British Honduras   
433           1  Unspecified                 ...  British  British Honduras   

            location locale        document_date     source  \
55      Philadelphia         1791-10-25 00:00:00  Newspaper   
126  Mosquito Coast          1777-02-24 00:00:00   Archival   
127   Mosquito Shore         1777-02-24 00:00:00    Archive   
129   Mosquito Shore         1777-02-24 00:00:00    Archive   
130   Mosquito Shore         1777-02-24 00:00:00    Archive   
..               ...    ...                  ...        ...   
429   Mosquito Coast         1777-08-20 00:00:00    Archive   
430   Mosquito Coast         1777-08-20 00:00:00    Archive   
431   Mosquito Coast         1777-08-20 00:00:00    Archive   
432   Mosquito Coast         1777-08-20 00:00:00    Archive   
433   Mosquito Coast         1777-08-20 00:00:00    Archive   

                    record                                           citation  \
55   Runaway advertisement  DISA00094; America's Historical Newspapers [Ma...   
126               Registry  DISA00123; “Return of the Registry of Indians ...   
127               Registry  DISA00123; “Return of the Registry of Indians ...   
129               Registry  DISA00123; “Return of the Registry of Indians ...   
130               Registry  DISA00123; “Return of the Registry of Indians ...   
..                     ...                                                ...   
429               Registry  Zotero ID: DISA00123, “Ret

In [91]:
disa.disa_id.values

array(['10-10002', '10-10003', '10-10004', '10-10005', '10-10006',
       '10-10007', '10-10008', '10-10009', '10-10010', '10-10011',
       '10-10012', '10-10013', '10-10014', '10-10015', '10-10016',
       '10-10017', '10-10018', '10-10019', '10-10020', '10-10021',
       '10-10022', '10-10023', '10-10024', '10-10025', '10-10026',
       '10-10027', '10-10028', '10-10029', '10-10030', '10-10031',
       '10-10032', '10-10033', '10-10034', '10-10035', '10-10036',
       '10-10037', '10-10038', '10-10039', '10-10040', '10-10041',
       '10-10042', '10-10043', '10-10044', '10-10045', '10-10046',
       '10-10047', '10-10048', '10-10049', '10-10050', '10-10051',
       '10-10052', '10-10053', '10-10054', '10-10055', '10-10056',
       '10-10057', '10-10058', '10-10059', '10-10060', '10-10061',
       '10-10062', '10-10063', '10-10064', '10-10065', '10-10066',
       '10-10067', '10-10068', '10-10069', '10-10070', '10-10071',
       '10-10072', '10-10073', '10-10074', '10-10075', '10-100

In [94]:
disa.nation.value_counts()

British          365
                  54
Spanish           28
British           22
American          12
USA                6
British Coast      2
Honduras           2
United States      1
French             1
Name: nation, dtype: int64